### PR TITLE
BGL: Doc & tests improvements

### PR DESCRIPTION
**Pull request by MaelRL**
## Summary of Changes

Rough changelog:

- Added BGL graph concepts (`IncidenceGraph`, etc.) to `General.txt` to get proper links everywhere (except for `Graph` to avoid accidental links).
- Clarified and rewrote many sentences of the BGL manual.
- Cleaned `\sa` and `\hasmodel` in the CGAL graph concepts (redirect to the dedicated pages).
- Fixed a few types in the BGL reference manual and made some pages with tables easier to read.
- Fixed a bunch of wrong "refines XxxGraph" (e.g. `HalfedgeGraph` refines `BidirectionalGraph` here and there) and added missing ones (models of `AdjacencyMatrix`, ...)
- Fixed T2 graph traits not being a proper model of the BGL/CGAL concepts (missing functions, missing types) and actually test the concept.
- Added null_vertex() to Arrangement_2's graph traits (`vertex_descriptor()`) and a test of arrangement. Currently it's only a model of the boost graph concepts, waiting for https://github.com/afabri/cgal/tree/Arrangement_2-graph_traits-GF.

## Release Management

* Affected package(s): `BGL`, `Triangulation_2`
* Issue(s) solved (if any): --
* Feature/Small Feature (if any): --

### CODE REVIEW

File: BGL/doc/BGL/BGL.txt [hunk 1]
```diff
@@ -158,51 +175,62 @@ See below a sample call of a function that uses the optional BGL named parameter
...
 \section BGLHeader Header Files, Namespaces, and Naming Conventions
 
-Partial specializations of the `boost::graph_traits<Graph>` for the \cgal package `PACKAGE` are in the
-namespace `boost` and in the header file `CGAL/boost/graph/graph_traits_PACKAGE.h`.   The
-free functions are in the namespace `CGAL`, as the compiler uses argument dependent lookup to find them.
-The %Euler operations are in the namespace `CGAL::Euler`, as the function `remove_face()` is at
-the same time a low level and an %Euler operation.
-Concerning the naming conventions we have to use those of \sc{Bgl}, as we have to fulfill the requirements of the concepts defined in \sc{Bgl}.
+This package provides the necessary classes and functions that enable using the
+some \cgal data structures as models of the \sc{Bgl} graph concepts.
```
Comment (sloriot) [inline BGL/doc/BGL/BGL.txt]: the some

File: BGL/doc/BGL/BGL.txt [hunk 2]
```diff
@@ -158,51 +175,62 @@ See below a sample call of a function that uses the optional BGL named parameter
...
+Free functions are in the namespace `CGAL`, and the compiler uses argument-dependent lookup to find them.
+%Euler operations, described in Section \ref BGLEulerOperations, are in the namespace `CGAL::Euler`, as the function `remove_face()` is at
+the same time a low-level and an %Euler operation.
+Concerning the naming conventions, we have to use those of the \sc{Bgl},
+as to fulfill the requirements of the concepts defined in the \sc{Bgl}.
+
+Note that these partial specializations are often providing more than
+is required, making these classes not only models of the graph concepts
+of the \sc{Bgl}, but also models of the CGAL graph concepts, that will
+described in detail in Section \ref BGLExtensions. Correspondence tables
```
Comment (sloriot) [inline BGL/doc/BGL/BGL.txt]: be

**Comment by sloriot** [comment]
What do you think of having a unique implementation for the 8 `graph_traits` headers in the `Triangulation_2` package? Shall I open an issue for it?

**Comment by MaelRL** [comment]
Merging the `Triangulation_2` graph traits is orthogonal to this PR, but it'd probably be a good thing to do. I only skimmed through those classes to add the missing parts, so I don't know if there are some tiny details that absolutely require having all of these classes rather than a common one (but I guess not). To be discussed with @afabri (in the issue).

**Comment by sloriot** [comment]
#2815 tracks that.

**Comment by MaelRL** [comment]
@afabri @sloriot Some of the named parameters (`vertex_point`, `halfedge_index`, etc.) are used with either with `CGAL::` or `boost::` (depending on the opinion of the author, I guess). It can sometimes create somewhat confusing code. Why are there two different possibilities and should that number be decreased to one (and which one!) ?

**Comment by afabri** [comment]
When we call a real BGL function such as  `boost::dijkstra()`, I would tend to use `boost::vertex_index`, but when we call a function dealing with a model of a `HalfedgeGraph` I would tend to use `CGAL::vertex_index`.

**Comment by sloriot** [comment]
@MaelRL's question seems to be more about extensions of CGAL, that is tags that cannot be used outside of CGAL. My personnal point of view if that everything that is added by CGAL should be in CGAL namespace to avoid confusion.

**Comment by afabri** [comment]
It would look strange to me to write   `great_function(mesh, (get(boost::vertex_index, mesh), get(boost::face_index, mesh))`.

**Comment by sloriot** [comment]
Then only the tags coming from boost should be in both namespaces and the extensions should only be in CGAL namespace.

**Comment by lrineau** [comment]
Do you all agree that we delay this PR integration to the release cycle for CGAL-4.13?

**Comment by afabri** [comment]
It globally improves things, so even if there is still room for improvement, why not 4.12 ?

**Comment by lrineau** [comment]
Because CGAL-4.12-beta1 is already 9 days late, and this PR was still never tested. That is why I would need a sound reason to force it into CGAL-4.12.

Other reasons for having a bad feeling for a short integration:
  - the patch is not short,
  - it modifies heavily the documentation, and for that reason I would like that authors of the BGL package (@afabri @sloriot) do a proofreading of the modified documentation.

**Comment by afabri** [comment]
ok

**Comment by afabri** [comment]
I have read the diff and it looks all correct to me.

**Comment by MaelRL** [comment]
Travis failing with the same error that master ( https://travis-ci.org/CGAL/cgal/builds/343362178 ) has.

Warning in the testsuite (can't resolve link to `Triangulation_2::Edge`) : can't reproduce locally, maybe a missing `CGAL::` throws it off.

**Comment by lrineau** [comment]
Doxygen version 1.8.4 detected a warning:
> /home/cgal-testsuite/cgal_doc_build/CGAL-4.12-Ic-179/doc/BGL/BGL.txt:67: Warning: unable to resolve link to `Triangulation_2::Edge' for \link command

What is strange is that Doxygen version 1.8.13 and later did not display any warning.

**Comment by MaelRL** [comment]
> Warning in the testsuite (can't resolve link to Triangulation_2::Edge) : can't reproduce locally, maybe a missing CGAL:: throws it off.

Can reproduce with 1.8.4 but I can't seem to fix it. Strangely, replacing the typedef with a function (`Triangulation_2::insert()`) will link properly, but none of the typedefs.

Can it be ignored?

**Comment by lrineau** [comment]
@sloriot You are our Doxygen expert. Let us know what we should do. As we plan to switch to Doxygen 1.8.13, maybe we can forget that error of Doxygen 1.8.4. But that warning will be annoying in the test results page. If we can remove it that would be simpler.

**Comment by lrineau** [comment]
This PR triggered new warnings in `test/BGL/`, on Windows:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-179.shtml#BGL

**Comment by lrineau** [comment]
Tested successfully in [CGAL-4.12-Ic-182], and then merged.

[CGAL-4.12-Ic-182]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.12-Ic-182.shtml

**Pull request merged**